### ISSUE TITLE
Infer return type of Union of functions when the functions are type stable

### ISSUE THREAD

**Issue by Seelengrab** [JuliaLang/julia#38068]
Referenced on slack by @brandondube

```
function g(bool_arg, arg)
   if bool_arg
     f = sin
   else
     f = cos
   end
   f(arg)
end
```

This function takes a boolean and applies either `sin` or `cos` to the other argument, depending on the value of the boolean. It allocates:

```
julia> @btime g(a, b) setup=(a=rand(Bool);b=1)
  27.739 ns (1 allocation: 16 bytes)
0.5403023058681398
```

and `@code_warntype` doesn't like the `Union{typeof(sin),typeof(cos)}` and infers `Any`:

```
julia> @code_warntype g(true, 1)
Variables
  #self#::Core.Compiler.Const(g, false)
  arg1::Bool
  arg2::Int64
  f::Union{typeof(cos), typeof(sin)}

Body::Any
1 ─      Core.NewvarNode(:(f))
└──      goto #3 if not arg1
2 ─      (f = Main.sin)
└──      goto #4
3 ─      (f = Main.cos)
4 ┄ %6 = (f)(arg2)::Any
└──      return %6
```

However, since we know that the second argument is `Int`, both methods have a single return type (and even share it!):

```
julia> Base.return_types(sin, (Int,))
1-element Array{Any,1}:
 Float64

julia> Base.return_types(cos, (Int,))
1-element Array{Any,1}:
 Float64
```

I suspect the allocation comes from the fact that the Any is inferred due to the Union, but because no matter which function is called `Float64` is returned, the compiler should be able to infer `Float64` for the Union here as well.

**Comment by Seelengrab** [comment]
Oh, totally forgot, this is 1.5.2.

**Comment by martinholters** [comment]
Looks fixed on master (presumably by #36623):
```julia
julia> VERSION
v"1.6.0-DEV.1331"

julia> @btime g(a, b) setup=(a=rand(Bool);b=1)
  17.001 ns (0 allocations: 0 bytes)
0.8414709848078965

julia> @code_warntype g(true, 1)
Variables
  #self#::Core.Const(g)
  bool_arg::Bool
  arg::Int64
  f::Union{typeof(cos), typeof(sin)}

Body::Float64
1 ─      Core.NewvarNode(:(f))
└──      goto #3 if not bool_arg
2 ─      (f = Main.sin)
└──      goto #4
3 ─      (f = Main.cos)
4 ┄ %6 = (f)(arg)::Float64
└──      return %6
```

**Comment by Seelengrab** [comment]
Yep, seems fixed, even for the more complicated case where the return types differ:

```
julia> function g(bool_arg, arg)
          if bool_arg
            f = sin
          else
            f = x -> "testreturn"
          end
          f(arg)
       end
g (generic function with 1 method)

julia> @btime g(a, b) setup=(a=rand(Bool);b=1)
  2.700 ns (0 allocations: 0 bytes)
"testreturn"

julia> @code_warntype g(true, 1)
Variables
  #self#::Core.Const(g)
  bool_arg::Bool
  arg::Int64
  #6::var"#6#7"
  f::Union{typeof(sin), var"#6#7"}

Body::Union{Float64, String}
1 ─      Core.NewvarNode(:(#6))
│        Core.NewvarNode(:(f))
└──      goto #3 if not bool_arg
2 ─      (f = Main.sin)
└──      goto #4
3 ─      (#6 = %new(Main.:(var"#6#7")))
└──      (f = #6)
4 ┄ %8 = (f)(arg)::Union{Float64, String}
└──      return %8
```
